---
title: "Estimation of multicellular factors from MOFA"
author:
- name: "Ricardo O. Ramirez Flores"
  affiliation: "Heidelberg University, Faculty of Medicine"
  email: "roramirezf@uni-heidelberg.de"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r, message=FALSE, warning=FALSE}
library(MOFA2)
library(tidyverse)
library(HDF5Array)
library(ComplexHeatmap)
library(scater)
library(scran)
library(uwot)
library(edgeR)
library(circlize)
```

```{r, echo=FALSE}
reticulate::use_condaenv(condaenv = "/Users/ricardoramirez/opt/miniconda3/envs/MOFA")
```

# Introduction

MOFA can be used to identify coordinated gene expression programs of multiple cell types from single cell RNA-seq data. Similarly as other computational tools such as scITD (link) or DIALOGUE (link), MOFA factors can be interpreted as shared gene programs between cell types, whose variability is associated to different covariates in complex experimental designs.

In this markdown we will estimate "multicellular factors" from a single-cell RNA-seq dataset of human myocardial infarction patients and healthy donors from a study by Kuppe, Ramirez Flores, Li, et al.

# Pseudobulk expression profiles and meta-data

In this section we import a SingleCellExperiment containing the profiles of 27 combinations of patient and physiological samples at different time points after myocardial infarction as well as control samples.

These specimens were classified in three major groups based on histological observations, nevertheless these are only approximations of a class

```{r}
# Defining the patient annotations -----------------------------------------------------------------
annotation_names <- tibble(patient_group = c("group_1", "group_2", "group_3"),
                           patient_group_name = c("myogenic-enriched", "ischemic-enriched", "fibrotic-enriched"))

batch_info <- read_csv("./data_MI/snrna_batch_ann.csv") %>%
  select(orig.ident, batch) %>%
  unique()

sample_dict <- read_csv("./data_MI/rna_patient_anns_revisions.csv",show_col_types = F) %>%
  left_join(annotation_names, by = "patient_group") %>%
  dplyr::select(-patient_group) %>%
  dplyr::rename("patient_group" = patient_group_name) %>%
  dplyr::mutate(patient_group = factor(patient_group, 
                                       levels = c("myogenic-enriched", "ischemic-enriched", "fibrotic-enriched"))) %>%
  #dplyr::select(-sample_id) %>%
  unique()

sample_dict_red <- sample_dict %>%
  left_join(batch_info, by = c("sample_id" = "orig.ident")) %>%
  dplyr::select(patient_region_id, major_labl, patient_group, batch) %>%
  unique() 
```

```{r, eval = F, echo= TRUE}
mi_pb <- loadHDF5SummarizedExperiment("./data_MI/integrated_rnasamples_ann_sce/")

ids <- colData(mi_pb) %>% 
  as.data.frame() %>% 
  left_join(sample_dict, by = c("orig.ident" = "sample_id")) %>%
  dplyr::select(cell_type, patient_region_id) %>%
  DataFrame()

mi_pb <- scuttle::sumCountsAcrossCells(x = mi_pb, 
                              ids  = ids)

assay(mi_pb, "counts") <- assay(mi_pb, "sum")
```

```{r, eval=F, echo=F}
saveRDS(mi_pb, "./data_MI/mi_pb.rds")
```

```{r, eval=T, echo=F}
mi_pb <- readRDS("./data_MI/mi_pb.rds")
```

Profiles with less than 10 cells are discarded

```{r}
ix <- which(colData(mi_pb)[,"ncells"] > 50)

mi_pb <- mi_pb[, ix]

mi_pb
```

We will now subset per major cell type

```{r}
ctypes <- set_names(colData(mi_pb)[,"cell_type"] %>%
                      unique()) 

mi_pb <- map(ctypes, function(ctype) { 
  
  ix <- which(colData(mi_pb)[,"cell_type"] == ctype)
  
  return(mi_pb[,ix])
  
  })

```

We will subset for cell-types of interest

```{r}
exclude_ct <- c("Mast", "Neuronal", "prolif", "Adipo")

mi_pb <- mi_pb[!names(mi_pb) %in% exclude_ct]
```

We will filter now genes with sufficient counts in at least 85% of the samples

```{r, warning=FALSE, message=FALSE}
mi_pb_red <- map(mi_pb, function(x) {
  
  useful_genes <- edgeR::filterByExpr(x, min.count = 10, min.prop = 0.85)

  return(x[useful_genes, ])
})
```

Standard log-normalization is performed

```{r}
mi_pb_red <- map(mi_pb_red, function(x) {
  
  return(scater::logNormCounts(x))
  
})
```

Per cell-type we will identify the top 5000 variable genes

```{r}
mi_pb_red <- map(mi_pb_red, function(x) {
  hvg <- getTopHVGs(x,var.threshold = 1)
  return(x[hvg, ])
})
```

Generate a long data-frame and prepare the MOFA model

```{r}
mi_pb_red <- map(mi_pb_red, function(x) {
  
  dat <- assay(x, "logcounts") 
  
  colnames(dat) <- colData(x)[,"patient_region_id"]
  
  dat %>%
    as.data.frame() %>%
    tibble::rownames_to_column("feature") %>%
    pivot_longer(-feature, names_to = "sample", values_to = "value")

}) %>% 
  enframe(name = "view") %>%
  unnest() %>%
  dplyr::mutate(feature = paste0(view, "_", feature))

saveRDS(mi_pb_red, "./data_MI/mi_pb_red.rds")
```

# Preparing and fitting the MOFA model 

We will try to identify 5 factors that explain the variability between patients captured by 7 different cell-types.

Every factor captures coordination of gene expression across cell types and will be called multicellular gene factors for the rest of the vignette.

It is important to clarify what these factors capture:

a) Coordinated expression of identical genes (generalistic response) across cell-types
b) Coordinated expression of different genes (cell-type specific response) across cell-types

At the same time this coordination could be associated to cell type location in the tissue, nevertheless this shouldn't be precisely true in global tissue responses.

```{r, message=FALSE}

MOFAobject <- create_mofa(mi_pb_red)

data_opts <- get_default_data_options(MOFAobject)

model_opts <- get_default_model_options(MOFAobject)

model_opts$num_factors <- 6

train_opts <- get_default_training_options(MOFAobject)

# Prepare MOFA model:
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)

# Train model:
outfile <- file.path("./results/MOFA_mcell/", "MI_model.hdf5")

model <- run_mofa(MOFAobject, outfile)
```

# Downstream analysis

## Association of factors to covariates

We will associate clinical covariates to factor scores. For each factor score, we will perform an analysis of variance to clinical covariates of interest. 

In this specific case we will identify which factors associate with the histological description

```{r}
# Associate with clinical data
factors <- get_factors(model, factors = "all")[[1]] %>%
  as.data.frame() %>%
  rownames_to_column("patient_region_id") %>%
  left_join(sample_dict_red, by = c("patient_region_id")) %>%
  pivot_longer(-colnames(sample_dict_red), names_to = "Factor")

predicted_label <- "patient_group"

# Get factors associated with patient group
factors <- factors %>%
  dplyr::select_at(c("patient_region_id", predicted_label, "Factor", "value")) %>%
  group_by(Factor) %>%
  nest() %>%
  mutate(pvalue = map(data, function(dat) {
    
    gene_aov <- aov(as.formula(paste0("value ~ ", predicted_label)), data = dat) %>%
      broom::tidy() %>%
      dplyr::filter(term == predicted_label) %>%
      dplyr::select(term, p.value)
    
    return(gene_aov)
  })) %>%
  tidyr::unnest(pvalue) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(adj_pvalue = p.adjust(p.value))
  
expl_var <- factors %>%
  dplyr::select(Factor, term, p.value, adj_pvalue)

expl_var
```

From this analysis we identify Factor 1 to be associated with the histological descriptions.

## Explained variance per factor and summary heatmap

We recover the explained variance per view per factor to evaluate if coordinated gene expression processes are coordinated between all the cell types or exclusive to only a group of them

```{r, fig.height=10, fig.width=8}
# Make the heatmap similar to scITD
row_anns <- get_factors(model, factors = "all")[[1]] %>% 
  rownames() %>% 
  enframe() %>% 
  dplyr::select(value) %>% 
  left_join(sample_dict_red, by = c("value" = "patient_region_id")) %>%
  dplyr::select(patient_group, major_labl)

row_ha = rowAnnotation(patient_group = row_anns$patient_group,
                       histology = row_anns$major_labl,
                       col = list(patient_group = c("fibrotic-enriched" = "blue",
                                                    "myogenic-enriched" = "red", 
                                                    "ischemic-enriched" = "darkgreen")))

# Add explain variance per cell-type
r2_per_factor <- model@cache$variance_explained$r2_per_factor[[1]]

column_ha = HeatmapAnnotation(r2 = r2_per_factor,assoc = -log10(expl_var$adj_pvalue))


scores_hmap <- Heatmap(get_factors(model, factors = "all")[[1]], 
        name = "factor_scores", 
        right_annotation = row_ha,
        top_annotation = column_ha,
        cluster_columns = FALSE, show_row_dend = FALSE)

scores_hmap
```

## Dimensionality reduction

Finally, to visualize in a 2D scatter plot we use UMAP with all factors

```{r}
# Factor scores per sample:
factors <- get_factors(model, factors = "all")[[1]] 

# UMAP of factor
factors_umap <- uwot::umap(factors, n_neighbors = 6)
colnames(factors_umap) <- c("UMAP_1", "UMAP_2")

factors_umap <- factors_umap %>%
  as.data.frame() %>%
  rownames_to_column("patient_region_id") %>%
  left_join(sample_dict_red)

umap_plt <- ggplot(factors_umap, aes(x = UMAP_1, 
                                     y = UMAP_2, 
                                     color = patient_group,
                                     shape = batch)) +
  geom_point(size = 2) +
  theme_classic() +
  theme(axis.text = element_text(size =12))

plot(umap_plt)
```

