---
title: "Grouping 2 datasets"
author: "Ricardo Ramirez"
date: "10/31/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r, message=FALSE, warning=FALSE}
library(MOFA2)
library(tidyverse)
library(HDF5Array)
library(ComplexHeatmap)
library(scater)
library(scran)
library(uwot)
library(edgeR)
library(circlize)
library(compositions)
library(biomaRt)
source("./MOFAcell/MOFAcellprep.R")
source("./MOFAcell/factorutils.R")
```

```{r, echo=FALSE}
reticulate::use_condaenv(condaenv = "/Users/ricardoramirez/opt/miniconda3/envs/MOFA")
```

# Introduction

MOFA can be used to identify coordinated gene expression programs of multiple cell types from single cell RNA-seq data. Similarly as other computational tools such as scITD (link) or DIALOGUE (link), MOFA factors can be interpreted as shared gene programs between cell types, whose variability is associated to different covariates in complex experimental designs.

In this markdown we will estimate "multicellular factors" from a single-cell RNA-seq dataset of HCM/ACM patients and healthy donors from a study by xxxxxx

# Pseudobulk expression profiles and meta-data

In this section we import a SingleCellExperiment containing the profiles of 27 combinations of patient and physiological samples at different time points after myocardial infarction as well as control samples.

These specimens were classified in three major groups based on histological observations, nevertheless these are only approximations of a class

```{r}
# Defining the patient annotations -----------------------------------------------------------------
sample_dict_red <- read_csv("./data_DCMACM_Science/meta_data.csv")[,-1]
# Disease DF
disease_dict <- sample_dict_red$disease %>% 
  unique() %>% 
  enframe(value = "disease", 
          name = "dis_code")

disease_dict$dis_code <- c("NF", "DCM", 
                           "ARVC", "NCC")

sample_dict_red <- sample_dict_red %>%
  left_join(disease_dict, by = "disease")

# Importing pb data
pb_data <- readRDS("./data_DCMACM_Science/pb_red_mat.rds")

# Importing coldata of the matrices
coldat <- read_csv("./data_DCMACM_Science/pb_coldata.csv",
                   show_col_types = FALSE)[,-1] %>%
  dplyr::mutate(cell_type = strsplit(colname,"_") %>%
                  map_chr(., ~ .x %>% last())) %>%
  column_to_rownames("colname") %>%
  dplyr::rename(ncells = "counts")

meta <- sample_dict_red %>%
  dplyr::rename("sample" = "donor_id")
```

# Now run MOFAcell

```{r, eval = T, echo= TRUE}
pb_dat <- SummarizedExperiment(assays = list("counts" = pb_data[,rownames(coldat)]), 
                               colData = DataFrame(coldat))
```

Finally prepare the data for the universal input of MOFAcell preprocessing

```{r, eval=T, echo=F}
counts <- assay(pb_dat, "counts")

coldata <- colData(pb_dat) %>%
  as.data.frame() %>%
  dplyr::rename("cell_counts" = "ncells") 

colnames(counts) <- rownames(coldata)
```

Generate a list of HVGs from another study

```{r}
broad_data <- readRDS("./data_HCMDCM_Nature/pb_red.rds")
```

Defining background genes

```{r}
mrkr_genes <- read_csv("./results/DCM_hubner/edgeR_cellmrkrs.csv")

mrkr_genes <- mrkr_genes %>% #dplyr::filter(!name %in% exclude_ct) %>%
  dplyr::filter(FDR < 0.01, logFC > 1) %>%
  dplyr::select(name, gene) %>%
  dplyr::rename("lineage" = name) %>%
  group_by(lineage) %>%
  nest() %>%
  dplyr::mutate(data = map(data, ~.x[[1]])) %>%
  deframe()
```


### First process counts data via MOFA cell prep

```{r}
# Define views of interest in ct
exclude_ct <- c("Mast", "Neuronal", "prolif", "Adipo", "none")

cts <- coldata$cell_type %>%
  unique()

cts <- cts[! cts %in% exclude_ct]

# Do all processing of counts matrices
pb_dat <- create_init_exp(counts = counts,
                          coldata = coldata) %>%
  filt_profiles(pb_dat = .,
                ncells = 25,
                cts = cts) %>%
  filt_gex_byexpr(pb_dat_list = .,
                  min.count = 10,
                  min.prop = 0.85) %>%
  tmm_trns(pb_dat_list = .) %>%
  filt_gex_byhvg(pb_dat_list = .,
                 prior_hvg = NULL,
                 var.threshold = 0) %>%
  filt_gex_bybckgrnd(pb_dat_list = ., 
                     prior_mrks = mrkr_genes) %>%
  pb_dat2MOFA(pb_dat_list = .)

pb_dat <- broad_data %>%
  dplyr::filter(feature %in% broad_data$feature)

saveRDS(pb_dat, "./data_DCMACM_Science/pb_red_wprior.rds")
```

# Preparing and fitting the MOFA model 

We will try to identify 5 factors that explain the variability between patients captured by 7 different cell-types.

Every factor captures coordination of gene expression across cell types and will be called multicellular gene factors for the rest of the vignette.

It is important to clarify what these factors capture:

a) Coordinated expression of identical genes (generalistic response) across cell-types
b) Coordinated expression of different genes (cell-type specific response) across cell-types

At the same time this coordination could be associated to cell type location in the tissue, nevertheless this shouldn't be precisely true in global tissue responses.

```{r, message=FALSE}

MOFAobject <- create_mofa(pb_dat)

data_opts <- get_default_data_options(MOFAobject)

model_opts <- get_default_model_options(MOFAobject)

model_opts$num_factors <- 6

train_opts <- get_default_training_options(MOFAobject)

# Prepare MOFA model:
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)

# Train model:
outfile <- file.path("./results/DCM_hubner/MOFA_mcell/", "DCM_hubner_wprior.hdf5")

model <- run_mofa(MOFAobject, outfile)
```

# Downstream analysis

## Association of factors to covariates

We will associate clinical covariates to factor scores. For each factor score, we will perform an analysis of variance to clinical covariates of interest. 

In this specific case we will identify which factors associate with the histological description

## Association of factors to covariates

We will associate clinical covariates to factor scores. For each factor score, we will perform an analysis of variance to clinical covariates of interest. 

In this specific case we will identify which factors associate with the histological description

```{r}
get_associated_factors <- function(meta, model, predicted_label) {
  
  factors <- get_allfscores(model = model, meta = meta)
  
  # Get factors associated with patient group
  factors <- factors %>%
  dplyr::select_at(c("sample", predicted_label, "Factor", "value")) %>%
  na.omit() %>%
  group_by(Factor) %>%
  nest() %>%
  mutate(pvalue = map(data, function(dat) {
    
    gene_aov <- aov(as.formula(paste0("value ~ ", predicted_label)), data = dat) %>%
      broom::tidy() %>%
      dplyr::filter(term == predicted_label) %>%
      dplyr::select(term, p.value)
    
    return(gene_aov)
  })) %>%
  tidyr::unnest(pvalue) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(adj_pvalue = p.adjust(p.value))

  expl_var <- factors %>%
  dplyr::select(Factor, term, p.value, adj_pvalue)

  return(expl_var)
  
}
```

```{r}
meta <- sample_dict_red %>%
  dplyr::rename("sample" = "donor_id")

expl_var_pgroup <- get_associated_factors(meta = meta,model = model,predicted_label = "disease")
expl_var_pgroup
```

```{r, fig.height=2.5, fig.width=4}
get_allfscores(model = model, meta = meta) %>%
  dplyr::filter(Factor %in% c("Factor1", "Factor2")) %>%
  pivot_wider(names_from = Factor, values_from = value) %>%
  ggplot(., aes(x = Factor1,
                y = Factor2,
                color = dis_code)) +
  geom_point(size = 2) +
  theme_classic() +
  theme(axis.text = element_text(size =12))
```

# Genetic variants

```{r}
expl_var_gen <- get_associated_factors(meta = meta %>%
                                            dplyr::filter(!Primary.Genetic.Diagnosis %in% c("control")),
                                                          model = model,
                                      predicted_label = "Primary.Genetic.Diagnosis")
expl_var_gen
```

```{r, fig.height=3, fig.width=4.5}
get_allfscores(model = model, meta = meta) %>%
  dplyr::filter(Factor %in% c("Factor1", "Factor2"),
                !Primary.Genetic.Diagnosis %in% c("control")) %>%
  pivot_wider(names_from = Factor, values_from = value) %>%
  ggplot(., aes(x = Primary.Genetic.Diagnosis,
                y = Factor2)) +
  geom_boxplot() +
  geom_point(size = 2) +
  theme_classic() +
  theme(axis.text = element_text(size =12),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust =0.5)) 
```

# Sex

```{r}
expl_var_sex <- get_associated_factors(meta = meta,
                                                          model = model,
                                      predicted_label = "sex")
expl_var_sex
```

From this analysis we identify Factor 1 to be associated with the disease descriptions.

## Explained variance per factor and summary heatmap

We recover the explained variance per view per factor to evaluate if coordinated gene expression processes are coordinated between all the cell types or exclusive to only a group of them

```{r, fig.height=10, fig.width=5}
ht_opt$ROW_ANNO_PADDING <- unit(2.5, "mm")
ht_opt$COLUMN_ANNO_PADDING <- unit(2.5, "mm")

# Association colors
col_fun_assoc <-  circlize::colorRamp2(seq(0, 20, length = 20), hcl.colors(20,"Purples",rev = T))
col_fun_r2 <- circlize::colorRamp2(seq(0, 100, length = 50), hcl.colors(50,"Oranges",rev = T))
col_fun_fact <- circlize::colorRamp2(seq(-5, 5, length = 50), hcl.colors(50,"Green-Brown",rev = T))

# Make the heatmap similar to scITD
row_anns <- get_factors(model, factors = "all")[[1]] %>% 
  rownames() %>% 
  enframe() %>% 
  dplyr::select(value) %>% 
  left_join(sample_dict_red, by = c("value" = "donor_id")) %>%
  dplyr::select(dis_code, sex)

row_ha <- rowAnnotation(patient_group = row_anns$dis_code,
                       sex = row_anns$sex,
                       col = list(patient_group = c("ARVC" = "lightblue",
                                                    "NF" = "darkred", 
                                                    "DCM"= "darkgreen",
                                                    "NCC" = "pink"),
                                  sex = c("male" = "black",
                                            "female" = "darkgrey")),
                       gap = unit(2.5, "mm"),
                       border = TRUE)


assoc_pvals <- bind_rows("disease" = -log10(expl_var_pgroup$adj_pvalue),
                         "genetics" = -log10(expl_var_gen$adj_pvalue),
                         "sex" = -log10(expl_var_sex$adj_pvalue)) %>%
  as.matrix()

# Add explain variance per cell-type
r2_per_factor <- model@cache$variance_explained$r2_per_factor[[1]]

column_ha <- HeatmapAnnotation("r2" = r2_per_factor,
                              "assocs" = assoc_pvals,
                              gap = unit(2.5, "mm"),
                              border = TRUE,
                              col = list(r2 = col_fun_r2,
                                  assocs = col_fun_assoc))


scores_hmap <- Heatmap(get_factors(model, factors = "all")[[1]], 
        name = "factor_scores", 
        right_annotation = row_ha,
        top_annotation = column_ha,
        cluster_columns = FALSE, 
        show_row_dend = TRUE,
        show_row_names = FALSE,
        border = TRUE,
        gap = unit(2.5, "mm"),
        col = col_fun_fact)

pdf("./results/DCM_hubner/MOFA_mcell/MOFAcell_summ_wprior.pdf", height = 5, width = 3.7)

draw(scores_hmap)

dev.off()

draw(scores_hmap)
```

## Paper stats

### Recovered mean across all samples
```{r}
model@cache$variance_explained$r2_total$single_group
model@cache$variance_explained$r2_total$single_group %>% mean()
```

### Factors associated with patient group
```{r}
useful_factors <- expl_var_pgroup %>%
  dplyr::filter(adj_pvalue < 0.15) %>%
  pull(Factor)

print(useful_factors)

model@cache$variance_explained$r2_per_factor$single_group[useful_factors,,drop = F] %>%
  colSums()

model@cache$variance_explained$r2_per_factor$single_group[useful_factors,,drop = F] %>%
  colSums() %>%
  mean()
```

### Reduced dimensions

```{r}
# Factor scores per sample:
factors <- get_factors(model, factors = "all")[[1]] 

# UMAP of factor
factors_umap <- uwot::umap(factors, n_neighbors = 6)
colnames(factors_umap) <- c("UMAP_1", "UMAP_2")

factors_umap <- factors_umap %>%
  as.data.frame() %>%
  rownames_to_column("donor_id") %>%
  left_join(sample_dict_red, by = "donor_id")

```

```{r}
umap_plt <- ggplot(factors_umap, aes(x = UMAP_1, 
                                     y = UMAP_2, 
                                     color = dis_code)) +
  geom_point(size = 2) +
  theme_classic() +
  theme(axis.text = element_text(size =12))

plot(umap_plt)
```

